designer/client/src/containers/hooks/useSearchQuery.tsx

```diff
@@ -1,26 +1,71 @@
-import { defaults } from "lodash";
+import { isEqual, omitBy, uniq, without } from "lodash";
 import * as queryString from "query-string";
-import { ParseOptions } from "query-string";
-import { useCallback, useMemo } from "react";
-import { defaultArrayFormat, setAndPreserveLocationParams } from "../../common/VisualizationUrl";
-import { UnknownRecord } from "../../types/common";
-import { useLocation, useNavigate } from "react-router-dom";
-
-export function useSearchQuery<T extends UnknownRecord>(options?: ParseOptions): [T, (v: T) => void] {
-    const navigate = useNavigate();
+import { ParsedQuery, ParseOptions } from "query-string";
+import { useCallback, useEffect, useMemo, useState } from "react";
+import { useLocation } from "react-router-dom";
+import { ensureArray } from "../../common/arrayUtils";
+
+type QueryRecord = ParsedQuery<string | boolean | number>;
+
+const DEFAULT_ARRAY_FORMAT: ParseOptions["arrayFormat"] = "comma";
+
+function parseQuery(searchString = window.location.search): QueryRecord {
+    return queryString.parse(searchString, {
+        arrayFormat: DEFAULT_ARRAY_FORMAT,
+        parseNumbers: true,
+        parseBooleans: true,
+        decode: true,
+    });
+}
+
+function stringifyQuery(params: QueryRecord): string {
+    const resultParams = omitBy(params, (value) => value === undefined || isEqual(value, []));
+    return queryString.stringify(resultParams, {
+        arrayFormat: DEFAULT_ARRAY_FORMAT,
+        encode: true,
+    });
+}
+
+export function useSearchQuery<T extends QueryRecord>(): [T, (v: T) => void] {
     const location = useLocation();
 
-    const query = useMemo(() => {
-        const parsedQuery = queryString.parse(location.search, defaults(options, { arrayFormat: defaultArrayFormat, parseBooleans: true }));
-        return parsedQuery as T;
-    }, [location.search, options]);
+    const [queryState, setQueryState] = useState(location.search);
+
+    const parsed = useMemo(() => parseQuery(queryState) as T, [queryState]);
 
-    const updateQuery = useCallback(
-        (value: T) => {
-            navigate({ search: setAndPreserveLocationParams(value) }, { replace: true });
-        },
-        [navigate],
+    const updateQuery = useCallback((value: T) => {
+        setQueryState(() => setAndPreserveLocationParams(value));
+    }, []);
+
+    useEffect(() => {
+        replaceSearchQuery(parsed);
+    }, [parsed]);
+
+    return [parsed, updateQuery];
+}
+
+export function replaceSearchQuery(params: QueryRecord): void {
+    const { history, location } = window;
+    const currentState = history.state || {};
+    const currentURL = new URL(location.href);
+    currentURL.search = stringifyQuery(params);
```

Comment:
Is it the reason to not use native UrlSearchParams here?

```
 const currentURL = new URL(location.href);
  const searchParams = new URLSearchParams(params);
  currentURL.search = searchParams.toString();
```

Reply:
we are using https://www.npmjs.com/package/query-string which gives us encode, decode, arrays etc. i just tried to unify and remove `navigate`.

Reply:
I even found similar doubts here https://dev.to/nerdyman/replacing-query-string-with-native-urlsearchparams-4kdg. It's not a blocker, so I can leave with this solution. You can resolve this comment if you think that there is no reason to change it.